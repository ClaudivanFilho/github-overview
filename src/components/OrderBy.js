import React, { Component } from 'react'
import PropTypes from 'prop-types'
import RepositoryContext from '../contexts/RepositoryContext'

const options = [{
  name: 'Name ASC',
  value: 'NAME_ASC',
  sort: (a, b) => a.nameWithOwner > b.nameWithOwner,
}, {
  name: 'Name DESC',
  value: 'NAME_DESC',
  sort: (a, b) => a.nameWithOwner < b.nameWithOwner,
}, {
  name: 'Issues ASC',
  value: 'ISSUES_ASC',
  sort: (a, b) => a.issues.totalCount > b.issues.totalCount,
}, {
  name: 'Issues DESC',
  value: 'ISSUES_DESC',
  sort: (a, b) => a.issues.totalCount < b.issues.totalCount,
}, {
  name: "PR's ASC",
  value: 'PRS_ASC',
  sort: (a, b) => a.pullRequests.totalCount > b.pullRequests.totalCount,
}, {
  name: "PR's DESC",
  value: 'PRS_DESC',
  sort: (a, b) => a.pullRequests.totalCount < b.pullRequests.totalCount,
}]

class OrderBy extends Component {
  static propTypes = {
    repositories: PropTypes.arrayOf(PropTypes.object),
    originalRepositories: PropTypes.arrayOf(PropTypes.object),
    updateRepositories: PropTypes.func,
  }

  doOrdenation = (opt) => {
    const repositories = [...this.props.repositories]
    const originalRepositories = [...this.props.originalRepositories]
    const selectedOption = options.find(option => option.value === opt)
    repositories.sort(selectedOption.sort)
    originalRepositories.sort(selectedOption.sort)
    this.props.updateRepositories(repositories, originalRepositories)
  }

  render() {
    return (
      <div className="flex items-center justify-end w-100 w-auto-m ma2 ma0-m">
        <select
          className="w-100 w4-m input-reset pl2 bg-white mid-gray b--silver ba br2 pointer f5"
          onChange={(opt) => {
            this.doOrdenation(opt.target.value)
          }}
        >
          {options.map(opt => (
            <option key={opt.value} value={opt.value}>
              {opt.name}
            </option>
          ))}
        </select>
      </div>
    )
  }
}

export default function ComponentWithContext(props) {
  return (
    <RepositoryContext.Consumer>
      {({ repositories, originalRepositories, updateRepositories }) => (
        <OrderBy
          {...props}
          repositories={repositories}
          updateRepositories={updateRepositories}
          originalRepositories={originalRepositories}
        />
      )}
    </RepositoryContext.Consumer>
  )
}

