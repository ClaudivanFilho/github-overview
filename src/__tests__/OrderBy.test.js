import React from 'react'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import OrderBy from '../components/OrderBy'

describe('Test OrderBy component', () => {
  let mountedComponent = null

  beforeAll(() => {
    mountedComponent = mount(
      <OrderBy />
    )
  })

  it('Renders correctly', () => {
    expect(toJson(mountedComponent)).toMatchSnapshot()
  })

  it('Component should be defined', () => {
    expect(mountedComponent).toBeDefined()
  })

  it('Must find a select component', () => {
    expect(mountedComponent.find('select').exists()).toBe(true)
  })

  it('Must find six options', () => {
    expect(mountedComponent.find('option')).toHaveLength(6)
  })
})
