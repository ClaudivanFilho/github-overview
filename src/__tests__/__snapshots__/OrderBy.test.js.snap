// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test OrderBy component Renders correctly 1`] = `
<ComponentWithContext>
  <OrderBy
    originalRepositories={Array []}
    repositories={Array []}
    updateRepositories={[Function]}
  >
    <div
      className="flex items-center"
    >
      <select
        className="pl2 bg-white br2 pointer"
        id="orderBy"
        onChange={[Function]}
      >
        <option
          key="NAME_ASC"
          value="NAME_ASC"
        >
          Name ASC
        </option>
        <option
          key="NAME_DESC"
          value="NAME_DESC"
        >
          Name DESC
        </option>
        <option
          key="ISSUES_ASC"
          value="ISSUES_ASC"
        >
          Issues ASC
        </option>
        <option
          key="ISSUES_DESC"
          value="ISSUES_DESC"
        >
          Issues DESC
        </option>
        <option
          key="PRS_ASC"
          value="PRS_ASC"
        >
          PR's ASC
        </option>
        <option
          key="PRS_DESC"
          value="PRS_DESC"
        >
          PR's DESC
        </option>
      </select>
    </div>
  </OrderBy>
</ComponentWithContext>
`;
